src/euler_one_qubit_decomposer.rs

```diff
@@ -10,15 +10,635 @@
+}
+
+#[pyclass]
+pub struct OneQubitGateErrorMap {
+    error_map: Vec<HashMap<String, f64>>,
+}
+
+#[pymethods]
+impl OneQubitGateErrorMap {
+    #[new]
+    fn new(num_qubits: Option<usize>) -> Self {
+        OneQubitGateErrorMap {
+            error_map: match num_qubits {
+                Some(n) => Vec::with_capacity(n),
+                None => Vec::new(),
+            },
+        }
+    }
+    fn add_qubit(&mut self, error_map: HashMap<String, f64>) {
+        self.error_map.push(error_map);
+    }
+}
+
+#[pyclass(sequence)]
+pub struct OneQubitGateSequence {
+    gates: Vec<(String, Vec<f64>)>,
+    #[pyo3(get)]
+    global_phase: f64,
+}
+
+#[pymethods]
+impl OneQubitGateSequence {
+    #[new]
+    fn new() -> Self {
+        OneQubitGateSequence {
+            gates: Vec::new(),
+            global_phase: 0.,
+        }
+    }
+    fn __getstate__(&self) -> (Vec<(String, Vec<f64>)>, f64) {
+        (self.gates.clone(), self.global_phase)
+    }
+
+    fn __setstate__(&mut self, state: (Vec<(String, Vec<f64>)>, f64)) {
+        self.gates = state.0;
+        self.global_phase = state.1;
+    }
+
+    fn __len__(&self) -> PyResult<usize> {
+        Ok(self.gates.len())
+    }
+
+    fn __getitem__(&self, py: Python, idx: SliceOrInt) -> PyResult<PyObject> {
+        match idx {
+            SliceOrInt::Slice(slc) => {
+                let len = self.gates.len().try_into().unwrap();
+                let indices = slc.indices(len)?;
+                let mut out_vec: Vec<(String, Vec<f64>)> = Vec::new();
+                // Start and stop will always be positive the slice api converts
+                // negatives to the index for example:
+                // list(range(5))[-1:-3:-1]
+                // will return start=4, stop=2, and step=-1
+                let mut pos: isize = indices.start;
+                let mut cond = if indices.step < 0 {
+                    pos > indices.stop
+                } else {
+                    pos < indices.stop
+                };
+                while cond {
+                    if pos < len as isize {
+                        out_vec.push(self.gates[pos as usize].clone());
+                    }
+                    pos += indices.step;
+                    if indices.step < 0 {
+                        cond = pos > indices.stop;
+                    } else {
+                        cond = pos < indices.stop;
+                    }
+                }
+                Ok(out_vec.into_py(py))
+            }
+            SliceOrInt::Int(idx) => {
+                let len = self.gates.len() as isize;
+                if idx >= len || idx < -len {
+                    Err(PyIndexError::new_err(format!("Invalid index, {idx}")))
+                } else if idx < 0 {
+                    let len = self.gates.len();
+                    Ok(self.gates[len - idx.unsigned_abs()].clone().into_py(py))
+                } else {
+                    Ok(self.gates[idx as usize].clone().into_py(py))
```

Comment:
I'm not sure I've fully understood PyO3, but I think `ToPyObject::to_object()` is meant for cases where you'd otherwise need `clone().into_py(py)`, since `IntoPy::into_py` takes `self`, whereas `ToPyObject::to_object()` takes `&self`.

Reply:
TBH, I don't remember the full context here,  I just copied this from what I did in rustworkx to add support for slices to the sequence types over there. I can give it a try to switch this to use `ToPyObject::to_object()` and see what happens.

Reply:
Yeah, I was able to do this and remove the explicit `clone()` in https://github.com/Qiskit/qiskit-terra/pull/9578/commits/37f3975286b06c7c3e55d9ad2ecc635f46cf93d2 so we only end up with a single copy now (going from rust to python) instead of 2.